doc/development/adding_templates.rst

```diff
@@ -6,75 +6,164 @@ Contributing templates
 
     import pennylane as qml
-    from pennylane.templates import template  # import the decorator
+    from pennylane.operation import Operation, AnyWires
 
-    @template
-    def MyNewTemplate(weights, wires):
+    class MyNewTemplate(Operation):
 
-        # Check that the inputs have the correct format
-        # ...
+        num_params = 1
+        num_wires = AnyWires
+        par_domain = "A"  # note: this attribute will be deprecated soon
 
-        for wire, weight in zip(wires, weights):
-            qml.RX(weight, wires=wire)
+        def expand(self):
 
-Since the template is called within a :ref:`quantum function <intro_vcirc_qfunc>`,
-it can only contain information processing that is allowed
-inside a quantum functions.
+            # extract the weights as the first parameter
+            # passed to this operation
+            weights = self.parameters[0]
 
-Make sure you consider the following:
+            # extract the length of the weights vector using
+            # the interface-agnostic `math` module
+            num_weights = qml.math.shape(weights)[0]
 
-* *Choose the name carefully.* Good names tell the user what a template is good for,
-  or what architecture it implements. To be consistent with other quantum operations
-  (which are classes), the function name (i.e., ``MyNewTemplate``) is written in camel case.
+            # record the ansatz in a tape
+            with qml.tape.QuantumTape() as tape:
 
-* *Register the template.* A template is "registered" by using the :mod:`template <pennylane.template>`
-  decorator ``@template``. This allows us to record the queue of operations of a template,
-  which is very useful for testing:
+                for i in range(num_weights):
+                    qml.RX(weights[i], wires=self.wires[i])
 
-  .. code-block:: python
+            return tape
 
-    with qml.utils.OperationRecorder() as rec:
-        MyNewTemplate(...)
 
-    list_of_gates = rec.queue
+The ``num_params`` and ``num_wires`` class attributes determine that an instance of this template can be created
+by passing a single parameter (of arbitrary shape and type), as well as an arbitrary number of wires:
 
-* Consider using the :func:`~.broadcast` function to make your
-  code more concise.
+.. code-block:: python
 
-* *Write an extensive docstring that explains how to use the template.* Include a sketch of the template (add the
-  file to the ``doc/_static/templates/<templ_type>/`` directory). You can also display a small usage example
-  at the beginning of the docstring.
+    weights = np.array([0.1, 0.2, 0.3])
+    MyNewTemplate(weights, wires=['a', 'b', 'd'])
+
+As an ``Operation``, templates can define other methods and attributes, such as a matrix representation,
+a generator, or even a gradient rule.
+
+.. note::
+
+    In principle, templates could also inherit from the :class:`Observable <pennylane.operation.Observable>`
+    class and define a sequence of diagonalising gates as an ansatz.
```

Comment:
🤔

Reply:
There is also nothing stopping a template that inherits from `Operation` from expanding to include measurements. I wonder what would happen, whether it would 'just work' or not.

Reply:
Can we leave it though? maybe it gives people ideas :)

Reply:
Yeah sure! The thinking face was more that it 'made me think' hahaha